### PR TITLE
feat(dop,pipeline): pipeline list support category filter

### PR DESCRIPTION
**Pull request by littlejiancc**
#### What this PR does / why we need it:
add category filter to pipeline

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMTQ2XSwiYXNzaWduZWUiOlsiMTAwMTI2MSJdfQ%3D%3D&id=292973&iterationID=1146&pId=292482&type=TASK)

#### Specified Reviewers:

/assign #Effet

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |     pipeline list support category filter         |
| 🇨🇳 中文    |       项目流水线分组从应用调整为“流水线分类”       |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: modules/dop/services/filetree/filetree.go [hunk 1]
```diff
@@ -192,6 +193,26 @@ func (svc *GittarFileTree) ListFileTreeNodes(req apistructs.UnifiedFileTreeNodeL
 						continue
 					}
 				}
+
+				// Filter does not match pipeline category rules
```
Comment (Effet) [inline modules/dop/services/filetree/filetree.go:197]: indent too deep, separate a func

**Review by Effet — CHANGES_REQUESTED** [review]

File: api/proto/core/pipeline/definition/definition.proto [hunk 2]
```diff
@@ -47,6 +47,11 @@ service DefinitionService {
       get: "/api/pipeline-definitions/actions/list-used-refs",
     };
   }
+  rpc StaticsGroupByFilePath (PipelineDefinitionStaticsRequest) returns (PipelineDefinitionStaticsResponse) {
+    option (google.api.http) = {
+      get: "/api/pipeline-definitions/actions/statics-group-by-filePath",
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply api/proto/core/pipeline/definition/definition.proto:52]: done

File: bundle/gittar.go [hunk 3]
```diff
@@ -382,7 +382,7 @@ func (b *Bundle) GetGittarBranchesV2(repo string, orgID string, onlyBranchNames
 }
 
 // 获取目录的子节点
-func (b *Bundle) GetGittarTreeNode(repo string, orgID string, simple bool, userID string) ([]apistructs.TreeEntry, error) {
+func (b *Bundle) GetGittarTreeNode(repo string, orgID string, simple bool, userID string) ([]apistructs.TreeEntry, string, error) {
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply bundle/gittar.go:385]: done

File: modules/dop/services/filetree/filetree.go (hunk 1 above)
Comment (littlejiancc) [reply modules/dop/services/filetree/filetree.go:197]: done

File: modules/dop/providers/projectpipeline/service_impl.go [hunk 4]
```diff
@@ -391,6 +392,13 @@ func (p *ProjectPipelineService) List(ctx context.Context, params deftype.Projec
 		Status:      params.Status,
 		AscCols:     params.AscCols,
 		DescCols:    params.DescCols,
+		IsOthers:    params.CategoryKey == apistructs.CategoryOthers,
+		FilePathWithNames: func() []string {
+			if params.CategoryKey != apistructs.CategoryOthers {
+				return apistructs.CategoryKeyRuleMap[apistructs.PipelineCategory(params.CategoryKey)]
+			}
+			return append(apistructs.CategoryKeyRuleMap[apistructs.CategoryBuildDeploy], apistructs.CategoryKeyRuleMap[apistructs.CategoryBuildArtifact]...)
+		}(),
```
Comment (Effet) [inline modules/dop/providers/projectpipeline/service_impl.go:401]: separate a func like: `getRulesByCategoryKey()`, and add unit test.

File: modules/dop/providers/projectpipeline/service_impl.go [hunk 5]
```diff
@@ -1351,3 +1359,92 @@ func (p *ProjectPipelineService) ListUsedRefs(ctx context.Context, params deftyp
 	}
 	return resp.Ref, nil
 }
+
+type pipelineCategoryRule struct {
```
Comment (Effet) [inline modules/dop/providers/projectpipeline/service_impl.go:1363]: I think `PipelineStatisticsByCategory` is more proper than `pipelineCategoryRule`

File: modules/dop/providers/projectpipeline/service_impl.go [hunk 6]
```diff
@@ -1351,3 +1359,92 @@ func (p *ProjectPipelineService) ListUsedRefs(ctx context.Context, params deftyp
...
+	Key        string
+	Category   string
+	Rules      []string
+	RunningNum uint64
+	FailedNum  uint64
+	TotalNum   uint64
```
Comment (Effet) [inline modules/dop/providers/projectpipeline/service_impl.go:1369]: you can separate these fields as a common struct like `PipelineStatisticsNums`

File: modules/dop/providers/projectpipeline/service_impl.go [hunk 7]
```diff
@@ -1351,3 +1359,92 @@ func (p *ProjectPipelineService) ListUsedRefs(ctx context.Context, params deftyp
...
+			OrgName:     org.Name,
+			ProjectName: project.Name,
+		}, cicdPipelineType),
+	})
+	if err != nil {
+		return nil, apierrors.ErrListProjectPipelineCategory.InternalError(err)
+	}
+	categoryRules := p.ListPipelineCategoryRule(ctx)
+
+	for _, statics := range staticsResp.PipelineDefinitionStatistics {
```
Comment (Effet) [inline modules/dop/providers/projectpipeline/service_impl.go:1417]: too complex, encapsulate a func like: `sumByCategory(key, allStatistics)`, and then
```
for i := range categories {
  r = rule[i]
  r.PipelineStatisticsNums = sumByCategory(r.key, defStatistics)
  ...
}
```

File: modules/pipeline/providers/definition/db/define.go [hunk 8]
```diff
@@ -249,13 +293,13 @@ func (client *Client) CountPipelineDefinition(req *pb.PipelineDefinitionListRequ
 }
 
 type PipelineDefinitionStatistics struct {
-	Remote     string
+	Group      string
 	FailedNum  uint64
 	RunningNum uint64
 	TotalNum   uint64
```
Comment (Effet) [inline modules/pipeline/providers/definition/db/define.go:299]: I suggest to separate a common struct like `PipelineStatisticsNums`

File: modules/pipeline/providers/definition/db/define.go [hunk 9]
```diff
@@ -249,13 +293,13 @@ func (client *Client) CountPipelineDefinition(req *pb.PipelineDefinitionListRequ
 }
 
 type PipelineDefinitionStatistics struct {
```
Comment (Effet) [inline modules/pipeline/providers/definition/db/define.go:295]: to unify a more proper name, I suggest `PipelineStatisticsByDefinition`

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/dop/providers/projectpipeline/service_impl.go (hunk 4 above)
Comment (littlejiancc) [reply modules/dop/providers/projectpipeline/service_impl.go:401]: done

File: modules/dop/providers/projectpipeline/service_impl.go (hunk 6 above)
Comment (littlejiancc) [reply modules/dop/providers/projectpipeline/service_impl.go:1369]: done

File: modules/dop/providers/projectpipeline/service_impl.go (hunk 7 above)
Comment (littlejiancc) [reply modules/dop/providers/projectpipeline/service_impl.go:1417]: done

**Review by Effet — APPROVED** [review]

**Pull request merged**